### PR TITLE
Feature deepspeech

### PR DESCRIPTION
**Pull request by PhilLecl**
Added support for [Mozilla's Project DeepSpeech](https://github.com/mozilla/DeepSpeech) offline speech-to-text-engine.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -21,6 +21,22 @@ stt_engines/snowboy/snowboydetect.py
...
+stt_engines/deepspeech/deepspeech-0.7.0-models.scorer
+
+stt_engines/deepspeech/deepspeech.log
+
+stt_engines/deepspeech/include/
+
+stt_engines/deepspeech/lib/
+
+stt_engines/deepspeech/share/
+
```
Comment (PhilLecl) [inline .gitignore]: That should be everything that's added automatically during the DeepSpeech installation or by running DeepSpeech

File: defaults/jarvis-config-default.sh [hunk 2]
```diff
@@ -3,6 +3,8 @@ command_stt="bing"
 conversation_mode=true
 dictionary="stt_engines/pocketsphinx/jarvis-dictionary.dic"
 gain="0"
+deepspeech_model="deepspeech-0.7.0-models.pbmm"
+deepspeech_scorer="deepspeech-0.7.0-models.scorer"
```
Comment (PhilLecl) [inline defaults/jarvis-config-default.sh]: Settings for which model and scorer DeepSpeech should use

File: stt_engines/deepspeech/main.sh [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+  jv_install python3 python3-pip virtualenv
+  cd "$(dirname "${BASH_SOURCE[0]}")"
+  virtualenv -p python3 .
+  source bin/activate
+  pip3 install deepspeech
+  deactivate
+  wget https://github.com/mozilla/DeepSpeech/releases/download/v0.7.0/deepspeech-0.7.0-models.pbmm
+  wget https://github.com/mozilla/DeepSpeech/releases/download/v0.7.0/deepspeech-0.7.0-models.scorer
+  cd "$jv_dir"
+}
```
Comment (PhilLecl) [inline stt_engines/deepspeech/main.sh]: - Make sure that Python3, pip3 and virtualenv are installed
- cd into "stt_engines/deepspeech"
- create a virtualenv (that uses Python3)
- activate said virtualenv
- install deepspeech via  pip
- deactivate the virtualenv
- download the model and scorer from Mozilla's GitHub repository
- cd into $jv_dir

File: stt_engines/deepspeech/main.sh [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+# Check if the needed files exist and install DeepSpeech if they don't
+[[ ( -f "$(dirname "${BASH_SOURCE[0]}")/deepspeech-0.7.0-models.pbmm" ) && \
+( -f "$(dirname "${BASH_SOURCE[0]}")/deepspeech-0.7.0-models.pbmm" ) && \
+( -d "$(dirname "${BASH_SOURCE[0]}")/bin" ) && \
+( -f "$(dirname "${BASH_SOURCE[0]}")/bin/deepspeech" ) ]] || {
```
Comment (PhilLecl) [inline stt_engines/deepspeech/main.sh]: Check if the DeepSpeech model, scorer and binary (and the bin directory of the virtualenv) exist

File: stt_engines/deepspeech/main.sh [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+  dialog_yesno "DeepSpeech doesn't seem to be installed.\nDo you want to install it?" true >/dev/null && {
+      stt_deepspeech_install
+      dialog_msg "DeepSpeech installed sucessfully"
+  }
```
Comment (PhilLecl) [inline stt_engines/deepspeech/main.sh:37]: Ask if DeepSpeech should be installed and run stt_deepspeech_install if yes

File: stt_engines/deepspeech/main.sh [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+}
+
+_deepspeech_transcribe () {
+  cd "$(dirname "${BASH_SOURCE[0]}")"
+  source bin/activate
+  transcribed="$(deepspeech --model $deepspeech_model $([[ -n "$deepspeech_scorer" ]] \
+    && echo "--scorer $deepspeech_scorer ")--audio $audiofile 2>"deepspeech.log")"
+  deactivate
+  cd "$jv_dir"
```
Comment (PhilLecl) [inline stt_engines/deepspeech/main.sh:55]: - cd into "stt_engines/deepspeech" and activate the virtualenv
- run deepspeech, store the transcript in $transcribed and write DeepSpeech's debug information into deepspeech.log
- deactivate the virtualenv and cd into $jv_dir

File: stt_engines/deepspeech/main.sh [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+  $verbose && jv_debug "DEBUG: $transcribed"
```
Comment (PhilLecl) [inline stt_engines/deepspeech/main.sh:57]: Thought I should add this, as stt_engines/google/main.sh also contains this line

File: stt_engines/deepspeech/main.sh [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+
+  echo $transcribed > $forder
```
Comment (PhilLecl) [inline stt_engines/deepspeech/main.sh:59]: Write the transcript into $forder

File: stt_engines/deepspeech/main.sh [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+}
+
+deepspeech_STT () { # STT () {} Listen & transcribes audio file then writes corresponding text in $forder
+    LISTEN $audiofile || return $?
+    _deepspeech_transcribe &
+   jv_spinner $!
+}
```
Comment (PhilLecl) [inline stt_engines/deepspeech/main.sh:66]: Essentially identical to the function in stt_engines/google/main.sh

File: stt_engines/deepspeech/uninstall.sh [hunk 10]
```diff
@@ -0,0 +1,11 @@
...
+
+cd "$(dirname "${BASH_SOURCE[0]}")"
+source bin/activate
+
+sudo pip uninstall deepspeech
+rm deepspeech*
+rm -rf bin include lib share
+
+deactivate
+cd "$jv_dir"
```
Comment (PhilLecl) [inline stt_engines/deepspeech/uninstall.sh]: - cd into stt_engines/deepspeech and activate the virtualenv
- uninstall deepspeech using pip
- remove the model, scorer and logfile
- remove the directories created by virtualenv
- deactivate the virtualenv and cd into $jv_dir

File: utils/configure.sh [hunk 11]
```diff
@@ -63,7 +63,7 @@ configure () {
                                    Weekly) check_updates=7;;
                                    Never)  check_updates=false;;
                                esac;;
-        command_stt)           options=('google' 'bing' 'wit' 'snowboy' 'pocketsphinx')
+        command_stt)           options=('google' 'bing' 'wit' 'snowboy' 'pocketsphinx' 'deepspeech')
```
Comment (PhilLecl) [inline utils/configure.sh:66]: Add deepspeech as an option in the command-stt-engine selection menu

**Review by PhilLecl — COMMENTED** [review]
Described the changes in detail

**Comment by alexylem** [comment]
Thanks a lot for this great contribution @PhilLecl  !
Quickly going through the changes looks good from coding structure standpoint.
@wikijm @Oliv4945 is one of you able to test it out before merging?

**Comment by Oliv4945** [comment]
Nice work! Unfortunately I have currently no functional test jig

**Pull request closed without merging**